test/e2e/cache/cache_test.go

```diff
@@ -0,0 +1,382 @@
+	test(cache)
+}
+
+func TestUpdateCacheCR(t *testing.T) {
+	t.Parallel()
+	ispn, cleanup := initCluster(t.Name(), false)
+	defer cleanup()
+	cacheName := ispn.Name
+	originalYaml := "localCache:\n  memory:\n    maxCount: 10\n"
+
+	// Create Cache CR with Yaml template
+	cr := cacheCR(cacheName, ispn)
+	cr.Spec.Template = originalYaml
+	testKube.Create(cr)
+	cr = testKube.WaitForCacheConditionReady(cacheName, tutils.Namespace)
+
+	validUpdateYaml := strings.Replace(cr.Spec.Template, "10", "50", 1)
+	cr.Spec.Template = validUpdateYaml
+	testKube.Update(cr)
+
+	// Assert CR spec.Template updated
+	testKube.WaitForCacheState(cacheName, tutils.Namespace, func(cache *v2alpha1.Cache) bool {
+		return cache.Spec.Template == validUpdateYaml
+	})
+
+	// Assert CR remains ready
+	cr = testKube.WaitForCacheConditionReady(cacheName, tutils.Namespace)
+
+	invalidUpdateYaml := `distributedCache: {}`
+	cr.Spec.Template = invalidUpdateYaml
+	testKube.Update(cr)
+
+	// Assert CR spec.Template updated
+	testKube.WaitForCacheState(cacheName, tutils.Namespace, func(cache *v2alpha1.Cache) bool {
+		return cache.Spec.Template == invalidUpdateYaml
+	})
+
+	// Wait for the Cache CR to become unready as the spec.Template cannot be reconciled with the server
+	testKube.WaitForCacheCondition(cacheName, tutils.Namespace, v2alpha1.CacheCondition{
+		Type:   v2alpha1.CacheConditionReady,
+		Status: metav1.ConditionFalse,
+	})
+}
+
+func TestCacheWithServerLifecycle(t *testing.T) {
+	t.Parallel()
+	ispn, cleanup := initCluster(t.Name(), true)
+	defer cleanup()
+
+	cacheName := ispn.Name
+	yamlTemplate := "localCache:\n  memory:\n    maxCount: \"%d\"\n"
+	originalConfig := fmt.Sprintf(yamlTemplate, 100)
+
+	// Create cache via REST
+	hostAddr, client := tutils.HTTPClientAndHost(ispn, testKube)
+	cacheHelper := tutils.NewCacheHelper(cacheName, hostAddr, client)
+	cacheHelper.CreateWithYaml(originalConfig)
+
+	// Assert CR created and ready
+	cr := testKube.WaitForCacheConditionReady(cacheName, tutils.Namespace)
+
+	// Assert that the owner reference has been correctly set to the Infinispan CR
+	if cr.GetOwnerReferences()[0].UID != ispn.UID {
+		panic("Cache has unexpected owner reference")
+	}
+
+	// Update cache configuration via REST
+	updatedConfig := fmt.Sprintf(yamlTemplate, 50)
+	cacheHelper.UpdateWithYaml(updatedConfig)
+
+	// Assert CR spec.Template updated
+	testKube.WaitForCacheState(cacheName, tutils.Namespace, func(cache *v2alpha1.Cache) bool {
+		return cache.Spec.Template == updatedConfig
+	})
+
+	// Delete cache via REST
+	cacheHelper.Delete()
+
+	// Assert CR deleted
+	err := wait.Poll(10*time.Millisecond, tutils.MaxWaitTimeout, func() (bool, error) {
+		return !testKube.AssertK8ResourceExists(cacheName, tutils.Namespace, &v2alpha1.Cache{}), nil
+	})
+	tutils.ExpectNoError(err)
+}
+
+func TestStaticServerCache(t *testing.T) {
```

Comment:
I've a failure for this test on OCP: the cache CR is not created and I have an error in the listener logs:
```
2021-12-03T16:13:16.374Z        ERROR   Error encountered for event 'cache-create': %!w(*fmt.wrapError=&{unable to CreateOrUpdate Cache CR: caches.infinispan.org "static-cache" is forbidden: cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on: , <nil> 0xc0004c3900})
```

Reply:
Good spot! This was because I was using `controllerutil.SetControllerReference` which is not valid outside of a controller. Instead it's necessary to use `controllerutil.SetOwnerReference`.